File: Source/Game/Podium/SoloPodium.cs
using System.Collections.Generic;
using WCSharp.Shared.Data;
using WCSharp.Api;
using System;

public static class SoloPodium
{
    private static Queue<(player Player, Point position)> PodiumQueue = new Queue<(player Player, Point position)>();
    private static List<unit> MovedUnits = new List<unit>();
    private static string PodiumType = "";
    private static string Color = Colors.COLOR_YELLOW;

    public static void BeginPodiumActions()
    {
        PodiumUtil.SetCameraToPodium();
        Utility.SimpleTimer(3.0f, ProcessPodiumTypeActions);
    }

    private static void EnqueueTopPlayerTimes()
    {
        var topTimes = PodiumUtil.SortPlayersFastestTime();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topTimes.Count - 1; i >= 0; i--)
        {
            var player = topTimes[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = "Time";
    }

    private static void EnqueueTopPlayerProgress()
    {
        var topProgress = PodiumUtil.SortPlayersTopProgress();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topProgress.Count - 1; i >= 0; i--)
        {
            var player = topProgress[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = "Progress";
    }

    private static void ProcessNextPodiumAction()
    {
        if (PodiumQueue.Count == 0)
        {
            PodiumUtil.EndingGameThankyou();
            return;
        }
        var (player, position) = PodiumQueue.Dequeue();
        var kitty = Globals.ALL_KITTIES[player].Unit;
        kitty.SetPosition(position.X, position.Y);
        kitty.SetFacing(270);
        kitty.IsPaused = true;
        MovedUnits.Add(kitty);
        Console.WriteLine($"{Colors.PlayerNameColored(player)}{Color} earned {PodiumUtil.PlacementString(PodiumQueue.Count + 1)} place for {PodiumType} with {GetStatBasedOnType(player)}|r");
        Utility.SimpleTimer(5.0f, ProcessNextPodiumAction);
    }


    private static void ProcessPodiumTypeActions()
    {
        PodiumQueue.Clear();
        PodiumUtil.ClearPodiumUnits(MovedUnits);
        switch (Gamemode.CurrentGameModeType)
        {
            case "Race":
                EnqueueTopPlayerTimes();
                break;
            case "Progression":
                EnqueueTopPlayerProgress();
                break;
        }
        ProcessNextPodiumAction();
    }

    private static string GetStatBasedOnType(player player)
    {
        var stats = Globals.ALL_KITTIES[player].TimeProg;
        switch (PodiumType)
        {
            case "Time":
                return stats.GetTotalTimeFormatted();
            case "Progress":
                return stats.GetOverallProgress().ToString("F2") + "%";
            default:
                return "n/a";
        }
    }

}

File: src/Game/Podium/SoloPodium.ts
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { Queue } from 'src/Utility/Queue'
import { Utility } from 'src/Utility/Utility'
import { MapPlayer, Point, Unit } from 'w3ts'
import { PodiumUtil } from './PodiumUtil'

export class SoloPodium {
    private static PodiumQueue = new Queue<[MapPlayer, Point]>()
    private static MovedUnits: Unit[] = []
    private static PodiumType: string = ''
    private static Color: string = Colors.COLOR_YELLOW_ORANGE
    private static Time: string = 'time'
    private static Progress: string = 'progress'

    public static BeginPodiumActions = () => {
        PodiumUtil.SetCameraToPodium()
        Utility.SimpleTimer(3.0, SoloPodium.ProcessPodiumTypeActions)
    }

    private static EnqueueTopPlayerTimes = () => {
        const topTimes = PodiumUtil.SortPlayersFastestTime()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topTimes.length - 1; i >= 0; i--) {
            const player = topTimes[i]
            const position = podiumPositions[i]
            SoloPodium.PodiumQueue.enqueue([player, position])
        }
        SoloPodium.PodiumType = SoloPodium.Time
    }

    private static EnqueueTopPlayerProgress = () => {
        const topProgress = PodiumUtil.SortPlayersTopProgress()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topProgress.length - 1; i >= 0; i--) {
            const player = topProgress[i]
            const position = podiumPositions[i]
            SoloPodium.PodiumQueue.enqueue([player, position])
        }
        SoloPodium.PodiumType = SoloPodium.Progress
    }

    private static ProcessNextPodiumAction = () => {
        if (SoloPodium.PodiumQueue.length === 0) {
            PodiumUtil.EndingGameThankyou()
            return
        }
        const [player, position] = SoloPodium.PodiumQueue.dequeue()!
        const kitty = Globals.ALL_KITTIES.get(player)!.Unit
        kitty.setPosition(position.x, position.y)
        kitty.setFacingEx(270)
        kitty.paused = true
        SoloPodium.MovedUnits.push(kitty)
        print(
            `${ColorUtils.PlayerNameColored(player)}${SoloPodium.Color} earned ${PodiumUtil.PlacementString(SoloPodium.PodiumQueue.length + 1)} for: place ${SoloPodium.PodiumType} with ${SoloPodium.GetStatBasedOnType(player)}|r`
        )
        Utility.SimpleTimer(5.0, SoloPodium.ProcessNextPodiumAction)
    }

    private static ProcessPodiumTypeActions = () => {
        SoloPodium.PodiumQueue.length = 0
        PodiumUtil.ClearPodiumUnits(SoloPodium.MovedUnits)
        switch (Globals.CurrentGameModeType) {
            case 'Race':
                SoloPodium.EnqueueTopPlayerTimes()
                break

            case 'Progression':
                SoloPodium.EnqueueTopPlayerProgress()
                break
        }
        SoloPodium.ProcessNextPodiumAction()
    }

    private static GetStatBasedOnType = (player: MapPlayer) => {
        const stats = Globals.ALL_KITTIES.get(player)!.TimeProg
        switch (SoloPodium.PodiumType) {
            case 'Time':
                return stats.GetTotalTimeFormatted()

            case 'Progress':
                return stats.GetOverallProgress().toFixed(2) + '%'

            default:
                return 'n/a'
        }
    }
}
